dsh_sdk/src/mqtt_token_fetcher.rs

```diff
@@ -0,0 +1,359 @@
+            .header("Authorization", authorization_header)
+            .json(payload)
+            .send()
+            .await;
+
+        match response {
+            Ok(response_success) => {
+                let status = response_success.status();
+                let body_text = response_success.text().await?;
+
+                match status {
+                    reqwest::StatusCode::OK => Ok(body_text),
+                    _ => Err(DshError::DshCallError {
+                        url: self.platform.endpoint_mqtt_token().to_string(),
+                        status_code: status,
+                        error_body: "Response NOT OK".to_string(),
+                    }),
+                }
+            }
+            Err(error) => Err(DshError::ReqwestError(error)),
+        }
+    }
+
+    fn hash_client_id(&self, id: &str) -> String {
+        let mut hasher = Sha256::new();
+        hasher.update(id);
+        let result = hasher.finalize();
+        format!("{:x}", result)
+    }
+}
+
+/// Represent Claims information for MQTT request
+/// * `action` - can be subscribe or publish
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Claims {
+    resource: Resource,
+    action: String,
+}
+
+impl Claims {
+    pub fn new(resource: Resource, action: String) -> Claims {
+        Claims { resource, action }
+    }
+}
+
+#[derive(derive_more::Display)]
+pub enum Actions {
+    Publish,
+    Subscribe,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct Resource {
+    stream: String,
+    prefix: String,
+    topic: String,
+    type_: Option<String>,
+}
+
+impl Resource {
+    pub fn new(stream: String, prefix: String, topic: String, type_: Option<String>) -> Resource {
+        Resource {
+            stream,
+            prefix,
+            topic,
+            type_,
+        }
+    }
+}
+
+/// Represents attributes associated with a mqtt token.
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "kebab-case")]
+struct MqttTokenAttributes {
+    gen: i32,
+    endpoint: String,
+    iss: String,
+    claims: Vec<Claims>,
+    exp: i32,
+    client_id: String,
+    iat: i32,
+    tenant_id: String,
+}
+
+/// Represents a mqtt token with its raw value and attributes.
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct MqttToken {
+    exp: i32,
+    raw_token: String,
+}
+
+impl MqttToken {
+    /// Creates a new instance of `MqttToken` from a raw token string.
+    ///
+    /// # Arguments
+    ///
+    /// * `raw_token` - The raw token string.
+    ///
+    /// # Returns
+    ///
+    /// A Result containing the created MqttToken or an error.
+    pub fn new(raw_token: String) -> Result<MqttToken, DshError> {
+        let header_payload = extract_header_and_payload(&raw_token)?;
+
+        let decoded_token = decode_payload(header_payload)?;
+
+        let token_attributes = MqttToken::parse_token_attributes(&decoded_token)?;
```

Comment:
parse_token fn can be replaced directly with 
```suggestion
        let token_attributes: MqttTokenAttributes = serde_json::from_slice(&decoded_token)?;
```